core/src/server/mod.rs

```diff
@@ -0,0 +1,324 @@
+        return Err((
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "Failed to send task".to_owned(),
+        ));
+    };
+
+    //Wait for the scheduler response
+    let (_, _, repo) = match task.rx.lock().unwrap().recv().unwrap() {
+        Ok(d) => d,
+        Err(err) => {
+            log::error!("Failed to scan data: {err}");
+            return Err((
+                StatusCode::INTERNAL_SERVER_ERROR,
+                "Failed to scan data".to_owned(),
+            ));
+        }
+    };
+
+    //Create the tarball
+    let tarball = format!("{}/{}.tar", &repo.scanner_path, &repo.folder_name);
+    let file = File::create(&tarball).unwrap();
+    let mut a = Builder::new(file);
+    a.append_file(
+        "converted.json",
+        &mut File::open(&repo.converted_file_path).unwrap(),
+    )
+    .unwrap();
+    a.append_file(
+        "extracted.json",
+        &mut File::open(&repo.extracted_file_path).unwrap(),
+    )
+    .unwrap();
+
+    //Create a Compressed tarball
+    let compressed_tarball = format!("{tarball}.gz");
+    let output_file = File::create(&compressed_tarball).unwrap();
+    let mut encoder = GzEncoder::new(output_file, Compression::default());
+    let mut builder = Builder::new(&mut encoder);
+    builder
+        .append_file(
+            "converted.json",
+            &mut File::open(&repo.converted_file_path).unwrap(),
+        )
+        .unwrap();
+    builder
+        .append_file(
+            "extracted.json",
+            &mut File::open(&repo.extracted_file_path).unwrap(),
+        )
+        .unwrap();
+    builder.finish().unwrap();
+
+    //Sending the tarball to the client
+    let file = match tokio::fs::File::open(&compressed_tarball).await {
+        Ok(file) => file,
+        Err(err) => {
+            log::error!("Failed to open file: {err}");
+            return Err((
+                StatusCode::INTERNAL_SERVER_ERROR,
+                "Failed to open file".to_owned(),
+            ));
+        }
+    };
+    let stream = ReaderStream::new(file);
+    let body = StreamBody::new(stream);
+    //Setting response headers
+    let mut headers = HeaderMap::new();
+    headers.insert(header::CONTENT_TYPE, "application/gzip".parse().unwrap());
+    headers.insert(header::CONTENT_ENCODING, "gzip".parse().unwrap());
+    //TODO: add a proper content length header
+    headers.insert(
+        header::CONTENT_DISPOSITION,
+        format!(
+            "attachment; filename={}",
+            format_args!("{}.tar.gz", repo.folder_name)
+        )
+        .parse()
+        .unwrap(),
+    );
+
+    //TODO: send back tarball of scanned data
+    Ok((StatusCode::OK, headers, body))
+}
+
+async fn get_extracted(
+    server: Extension<Arc<Server>>,
+    Json(payload): Json<ScanRequest>,
+) -> impl IntoResponse {
+    let repo = payload.repo_url.unwrap_or_default();
+    let conf = crate::config::Config::new();
+    let task = Arc::new(Task::new(scan, Arc::new(repo), Arc::new(conf)));
```

Comment:
The idea is that when a client request the `converted` data from the server the first time, we cache the `scanned` data and return them only they wants. If they or another user make a second to the same repository but now requesting `extracted` data we won't have to generate the extracted data a second time and it will be rather fast.